src/Network/WebSockets/Snap.hs

```diff
@@ -120,37 +62,36 @@ runWebSocketsSnap = runWebSocketsSnapWith WS.defaultConnectionOptions
 --------------------------------------------------------------------------------
 -- | Variant of 'runWebSocketsSnap' which allows custom options
 runWebSocketsSnapWith
-    :: Snap.MonadSnap m
-    => WS.ConnectionOptions
-    -> WS.ServerApp
-    -> m ()
+  :: Snap.MonadSnap m
+  => WS.ConnectionOptions
+  -> WS.ServerApp
+  -> m ()
 runWebSocketsSnapWith options app = do
-    rq <- Snap.getRequest
-    Snap.escapeHttp $ \tickle writeEnd -> do
-
-        thisThread <- lift myThreadId
-        mvar       <- lift newEmptyMVar
-        parse      <- lift $ copyMVarToStream mvar
-        write      <- lift $ copyStreamToIteratee writeEnd
-        stream     <- lift $ WS.makeStream parse write
-
-        let options' = options
-                    { WS.connectionOnPong = do
-                            tickle (max 60)
-                            WS.connectionOnPong options
-                    }
-
-            pc = WS.PendingConnection
-                    { WS.pendingOptions  = options'
-                    , WS.pendingRequest  = fromSnapRequest rq
-                    , WS.pendingOnAccept = forkPingThread tickle
-                    , WS.pendingStream   = stream
-                    }
-
-        _ <- lift $ forkIO $ finally (app pc) $ do
-            WS.close stream
-            throwTo thisThread ServerAppDone
-        copyIterateeToMVar tickle mvar
+  rq <- Snap.getRequest
+  Snap.escapeHttp $ \tickle readEnd writeEnd -> do
+    readEnd'  <- Streams.lockingInputStream  readEnd
+    writeEnd' <- Streams.lockingOutputStream writeEnd
```

Comment:
I am not sure if locking these is necessary.  `websockets` already adds locking in `makeStream`, so I'm worried the double-locking might impact performance a bit.